packages/back-end/src/services/experimentTimeSeries.ts

```diff
@@ -0,0 +1,232 @@
+    })
+  );
+
+  await context.models.metricTimeSeries.upsertMultipleSingleDataPoint(
+    metricTimeSeriesSingleDataPoints
+  );
+}
+
+function convertMetricToMetricValue(
+  metric: SnapshotMetric | undefined
+): MetricTimeSeriesValue | undefined {
+  if (!metric) {
+    return undefined;
+  }
+
+  // NB: Explicitly naming all fields to benefit from type safety
+  // when SnapshotMetric and MetricTimeSeriesDataPoint change
+  return {
+    value: metric.value,
+    denominator: metric.denominator,
+    expected: metric.expected,
+    ci: metric.ci,
+    pValue: metric.pValue,
+    pValueAdjusted: metric.pValueAdjusted,
+    chanceToWin: metric.chanceToWin,
+  };
+}
+
+const hashObject = (obj: object) => md5(JSON.stringify(obj));
+
+function getExperimentSettingsHash(
+  snapshotSettings: ExperimentSnapshotSettings,
+  snapshotAnalysisSettings: ExperimentSnapshotAnalysisSettings
+): string {
+  return hashObject({
+    // snapshotSettings
+    activationMetric: snapshotSettings.activationMetric,
+    attributionModel: snapshotSettings.attributionModel,
+    queryFilter: snapshotSettings.queryFilter,
+    segment: snapshotSettings.segment,
+    skipPartialData: snapshotSettings.skipPartialData,
+    datasourceId: snapshotSettings.datasourceId,
+    exposureQueryId: snapshotSettings.exposureQueryId,
+    startDate: snapshotSettings.startDate,
+    regressionAdjustmentEnabled: snapshotSettings.regressionAdjustmentEnabled,
+    experimentId: snapshotSettings.experimentId,
+
+    // analysisSettings
+    dimensions: snapshotAnalysisSettings.dimensions,
+    statsEngine: snapshotAnalysisSettings.statsEngine,
+    regressionAdjusted: snapshotAnalysisSettings.regressionAdjusted,
+    sequentialTesting: snapshotAnalysisSettings.sequentialTesting,
+    sequentialTestingTuningParameter:
+      snapshotAnalysisSettings.sequentialTestingTuningParameter,
+    baselineVariationIndex: snapshotAnalysisSettings.baselineVariationIndex,
+    pValueCorrection: snapshotAnalysisSettings.pValueCorrection,
+  });
+}
+
+function getMetricSettingsHash(
+  metricId: string,
+  metricSettings: MetricForSnapshot,
+  factMetrics?: FactMetricInterface[],
+  factTableMap?: Map<string, FactTableInterface>
+): string {
+  const factMetric = factMetrics?.find((metric) => metric.id === metricId);
+  if (!factMetric) {
+    return hashObject(metricSettings);
+  } else {
+    const numeratorFactTableId = factMetric.numerator.factTableId;
+    const numeratorFactTable = numeratorFactTableId
+      ? factTableMap?.get(numeratorFactTableId)
+      : undefined;
+
+    const denominatorFactTableId = factMetric.denominator?.factTableId;
+    const denominatorFactTable = denominatorFactTableId
+      ? factTableMap?.get(denominatorFactTableId)
+      : undefined;
+
+    const numeratorFilters = numeratorFactTable?.filters.filter((it) =>
+      factMetric.numerator.filters.includes(it.id)
+    );
+
+    // TODO: Some of these have a `updatedate`, should we include it or filter it out?
```

Comment:
I'm not sure which you mean, but if they are on the fact table, I would filter it out to be a bit "riskier" but with fewer false positives.